site/source/docs/porting/simd.rst

```diff
@@ -3,27 +3,76 @@
+5. Compile existing SIMD code that uses the ARM NEON intrinsics (``#include <arm_neon.h>``)
+
+These techniques can be freely combined in a single program.
+
+To enable any of the five types of SIMD above, pass the WebAssembly-specific ``-msimd128`` flag at compile time. This will also turn on LLVM's autovectorization passes. If that is not desirable, additionally pass flags ``-fno-vectorize -fno-slp-vectorize`` to disable the autovectorizer. See `Auto-Vectorization in LLVM <https://llvm.org/docs/Vectorizers.html>`_ for more information.
+
+WebAssembly SIMD is supported by
+
+* Chrome ≥ 91 (May 2021),
+
+* Firefox ≥ 89 (June 2021),
+
+* Safari ≥ 16.4 (March 2023) and
+
+* Node.js ≥ 16.4 (June 2021).
 
-Emscripten supports the `WebAssembly SIMD proposal <https://github.com/webassembly/simd/>`_ when using the WebAssembly LLVM backend. To enable SIMD, pass the -msimd128 flag at compile time. This will also turn on LLVM's autovectorization passes, so no source modifications are necessary to benefit from SIMD.
+See `WebAssembly Roadmap <https://webassembly.org/roadmap/>`_ for details about other VMs.
 
-At the source level, the GCC/Clang `SIMD Vector Extensions <https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html>`_ can be used and will be lowered to WebAssembly SIMD instructions where possible. In addition, there is a portable intrinsics header file that can be used.
+An upcoming `Relaxed SIMD proposal <https://github.com/WebAssembly/relaxed-simd/tree/main/proposals/relaxed-simd>`_ will add more SIMD instructions to WebAssembly.
+
+================================
+GCC/Clang SIMD Vector Extensions
+================================
+
+At the source level, the GCC/Clang `SIMD Vector Extensions <https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html>`_ can be used and will be lowered to WebAssembly SIMD instructions where possible.
+
+This enables developers to create custom wide vector types via typedefs, and use arithmetic operators (+,-,*,/) on the vectorized types, as well as allow individual lane access via the vector[i] notation. However, the `GCC vector built-in functions <https://gcc.gnu.org/onlinedocs/gcc/x86-Built-in-Functions.html>`_ are not available. Instead, use the WebAssembly SIMD Intrinsics functions below.
+
+===========================
+WebAssembly SIMD Intrinsics
+===========================
+
+LLVM maintains a WebAssembly SIMD Intrinsics header file that is provided with Emscripten, and adds type definitions for the different supported vector types.
 
     .. code-block:: cpp
 
        #include <wasm_simd128.h>
+       #include <stdio.h>
+
+       int main() {
+       #ifdef __wasm_simd128__
+         __f32x4 v1 = wasm_f32x4_make(1.2f, 3.4f, 5.6f, 7.8f);
+         __f32x4 v2 = wasm_f32x4_make(2.1f, 4.3f, 6.5f, 8.7f);
+         __f32x4 v3 = v1 + v2;
+         // Prints "v3: [3.3, 7.7, 12.1, 16.5]"
+         printf("v3: [%.1f, %.1f, %.1f, %.1f]\n", v3[0], v3[1], v3[2], v3[3]);
```

Comment:
Similarly, the result of the indexing operation depends on internal implementation details in the header, so it is best not to use it.
```suggestion
         printf("v3: [%.1f, %.1f, %.1f, %.1f]\n",
                wasm_f32x4_extract_lane(v3, 0),
                wasm_f32x4_extract_lane(v3, 1),
                wasm_f32x4_extract_lane(v3, 2),
                wasm_f32x4_extract_lane(v3, 3));
```

Reply:
Is there a way to disallow the [] notation from compiling, if it is unreliable?

Reply:
No, unfortunately not that I can think of. C simply lacks the kind of information-hiding we would need to make the internal representation of v128_t unobservable.